internal/pkg/agent/application/upgrade/step_download.go

```diff
@@ -119,3 +117,50 @@ func newVerifier(version string, log *logger.Logger, settings *artifact.Config)
 
 	return composed.NewVerifier(fsVerifier, snapshotVerifier, remoteVerifier), nil
 }
+
+func (u *Upgrader) downloadWithRetries(
+	ctx context.Context,
+	downloaderCtor func(string, *logger.Logger, *artifact.Config) (download.Downloader, error),
+	version string,
+	settings *artifact.Config,
+) (string, error) {
+	ctx, cancel := context.WithTimeout(ctx, u.settings.Timeout)
+	defer cancel()
+
+	expBo := backoff.NewExponentialBackOff()
+	expBo.InitialInterval = settings.RetrySleepInitDuration
+
+	boMaxRetries := backoff.WithMaxRetries(expBo, uint64(settings.RetryMaxCount))
+	boCtx := backoff.WithContext(boMaxRetries, ctx)
+
+	var path string
+	var attempt uint
+
+	opFn := func() error {
+		attempt++
+		u.log.Debugf("download attempt %d of %d", attempt, settings.RetryMaxCount+1)
+
+		downloader, err := downloaderCtor(version, u.log, settings)
```

Comment:
Unfortunately, yes.  Initially I had left this out of the retry loop but then I realized that the downloader constructor ([`newDownloader`](https://github.com/elastic/elastic-agent/blob/8b8b3209ed02d2832c809e5cdd645506dacdc866/internal/pkg/agent/application/upgrade/step_download.go#L80-L97)) eventually ends up making [a call ](https://github.com/elastic/elastic-agent/blob/8b8b3209ed02d2832c809e5cdd645506dacdc866/internal/pkg/agent/application/upgrade/artifact/download/snapshot/downloader.go#L97-L98)to `https://artifacts-api.elastic.co/v1/search/...` in case we're upgrading to `-SNAPSHOT` version artifacts.  Since that network call could fail transiently, I had to recreate downloader inside the retry loop.

Reply:
ah yeah for snapshots